### PR TITLE
Test for mitgcm format error

### LINKED ISSUE #20: MITgcm error when reading in float file
I am encountering an error when running the MITgcm that appears in the .e file for the job I submitted to my cluster (not in STDERR or STDOUT):
```
In source file mdsio_rd_rec_rl.f, at line number 1824
 at line number 1824
 File name = flt_CM2p6_global_hex.bin    unformatted, direct access   record = 37172161
 In source file mdsio_rd_rec_rl.f, at line number 1824
PGFIO-F-253/unformatted read/unit=9/attempt to read non-existent record (direct access).
 File name = flt_CM2p6_global_hex.bin    unformatted
```
The relevant section with line 1824 is shown below: 
 ```fortran
      IF ( debugLevel.GE.debLevC ) THEN
        WRITE(msgBuf,'(A,I9,2x,I9)')
     &      ' MDS_RD_REC_RL: iRec,Dim = ', iRec, nArr
        CALL PRINT_MESSAGE( msgBuf, standardMessageUnit,
     &                      SQUEEZE_RIGHT , myThid )
      ENDIF

      IF ( fPrec.EQ.precFloat32 ) THEN
        READ( dUnit, rec=iRec ) r4Buf <--------------------------------- LINE 1824 
        DO k=1,nArr
          arr(k) = r4Buf(k)
        ENDDO
      ELSEIF ( fPrec.EQ.precFloat64 ) THEN
        READ( dUnit, rec=iRec ) r8Buf
        DO k=1,nArr
          arr(k) = r8Buf(k)
        ENDDO
      ELSE
        WRITE(msgBuf,'(A,I9)')
     &        ' MDS_RD_REC_RL: illegal value for fPrec=',fPrec
        CALL PRINT_ERROR( msgBuf, myThid )
        STOP 'ABNORMAL END: S/R MDS_RD_REC_RL'
      ENDIF

      RETURN
      END
```
The float file I am attempting to use was generated with
```python 

fs = generators.FloatSet((-280,80), (-81,65.0), dx=.03125, dy=.03125)
model_grid = {'lon': mask_lon, 'lat': mask_lat, 'land_mask': mask_bath}
fs.to_mitgcm_format("flt_CM2p6_global_hex", model_grid=model_grid)
```
From visually inspection, I am fairly confident that the hexagonal tiling step is done correctly. However, given the recent changes to `to_mitgcm_format()` it is possible that the output formatting was broken. To test this, I will try to rebuild a float file with the new code that I know was previously working fine with the MITgcm.

### PR DESCRIPTION
**Pull request by rabernat**
Should kind of fix #20, to the extent possible.

It's not possible to output a large float set for MITgcm using 32-bit precision because we can't write accurately write the number of floats in the file header.

### CODE REVIEW

**Comment by rabernat** [comment]
Now we have an option to specify `read_binary_prec` in `to_mitgcm_format()`. This will be necessary for large float sets, since I guess we learned we can't actually represent large integers with 32 bits of precision. Kind of a bummer, because the KDTree masking step is *much* slower with 64 bit data. (So slow that I disabled the running of those tests on travis; but you can run them locally.)

**Comment by nathanieltarshish** [comment]
If the `kDTree` step is much slower with 64bit, then could we make just the writing to `fltmatrix` step be 64bit but keep the rest of the alogrithm unchanged (i.e. with normal Python type inferencing like before)? I'm not sure why we are altering the typing before the `fltmatrix` output step. Would you please clarify this?

**Comment by rabernat** [comment]
> I'm not sure why we are altering the typing before the fltmatrix output step. @rabernat would you please clarify this?

The code for hexmesh and rectmesh is unchanged with this PR. The datatype for those routines is set by the default output of `meshgrid`, which is float64. So contrary to what I said, it cant be the KDtree step that is causing the change in speed.

I ran some more detailed checking on the test function, and it looks like the difference in speed just comes from the cost of reading and writing a big (2 GB0 file over and over again. (It gets written and read three times per test.) At least on my latptop, this far exceeds the time for the KDTree.

```python
%time fs.get_rectmesh()
>>> CPU times: user 1min 12s, sys: 5.06 s, total: 1min 17s
    Wall time: 32.7 s
%time tg._actually_do_mitgcm_check(fs, tmpdir, prec=64)
>>> CPU times: user 15min 3s, sys: 1min 23s, total: 16min 26s
    Wall time: 7min 5s
```

Sorry for causing confusion.

**Comment by nathanieltarshish** [comment]
@rabernat Thanks for the clarification and explanation. Will we just run tests with ` prec=32` then? Or do you think we should split the tests up and also run a small domain with `prec=64`?

**Comment by rabernat** [comment]
Currently the only time `prec=64` is used in the tests is for a the example you came up with involving the very large domain. But since it takes 15 minutes, I don't have it run on travis.

File: floater/test/test_generators.py [hunk 1]
```diff
@@ -133,9 +157,28 @@ def test_area(fs_all):
...
+                    reason="Skipping this test on Travis CI because it's too slow.")
+@pytest.mark.xfail(reason="number of floats too big to represent with 32-bit float")
+def test_big_domain_32bit(fs_big, tmpdir):
+    _actually_do_mitgcm_check(fs_big, tmpdir, prec=32)
+
+
+@pytest.mark.skipif("TRAVIS" in os.environ and os.environ["TRAVIS"] == "true",
+                    reason="Skipping this test on Travis CI because it's too slow.")
+def test_big_domain_64bit(fs_big, tmpdir):
+    _actually_do_mitgcm_check(fs_big, tmpdir, prec=64)
```
Comment (rabernat) [inline floater/test/test_generators.py]: this is the case I am talking about

Comment (nathanieltarshish) [inline floater/test/test_generators.py]: Oh, I see. `@pytest.mark.skipif` is a cool feature.

**Comment by rabernat** [comment]
I think I will first merge #26, rebase this, and then merge this. Ok with you?

**Comment by nathanieltarshish** [comment]
One quick question before you merge: my funky big domain test has 29,490,575 floats, but now that I know what I am looking for (odd number of floats above 2^(24)+1), I can presumably find a smaller domain and faster example.  Do you think that is worth doing?

**Comment by rabernat** [comment]
> Do you think that is worth doing?

Definitely! It would be great to make that stupid test run faster. All you have to do is edit the ``fs_big`` function in `test_generators.py`. But it's probably easiest if I merge this first and you make that change via a new PR.

**Comment by nathanieltarshish** [comment]
Sounds good, will do!

**Pull request merged**